// @trimble-oss/modus-web-components/dist/collection/components/modus-navbar/profile-menu/modus-navbar-profile-menu.js
// eslint-disable-next-line
import { h } from '@stencil/core';
export class ModusNavbarProfileMenu {
  constructor() {
    this.avatarUrl = undefined;
    this.email = undefined;
    this.initials = undefined;
    this.links = undefined;
    this.reverse = undefined;
    this.username = undefined;
    this.variant = 'default';
  }
  render() {
    var _a;
    const direction = this.reverse ? 'reverse' : '';
    const variant = this.variant === 'default' ? '' : `profile-menu-${this.variant}`;
    return (h("div", { class: `profile-menu ${direction} ${variant}`, onClick: (event) => event.preventDefault() }, h("div", { class: "user" }, this.avatarUrl ? (h("img", { class: "avatar", src: this.avatarUrl, alt: "Avatar" })) : (h("span", { class: "initials" }, this.initials)), h("div", null, h("div", { class: "username" }, this.username), h("div", { class: "email" }, this.email))), ((_a = this.links) === null || _a === void 0 ? void 0 : _a.length) ? (h("div", { class: "links" }, this.links.map((link) => {
      return (h("div", { class: "link", onClick: () => this.linkClick.emit(link.id) }, link.display));
    }))) : null, h("div", { class: "sign-out", onClick: () => this.signOutClick.emit() }, h("div", null, "Sign out"))));
  }
  static get is() { return "modus-navbar-profile-menu"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["modus-navbar-profile-menu.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-navbar-profile-menu.css"]
    };
  }
  static get properties() {
    return {
      "avatarUrl": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "avatar-url",
        "reflect": false
      },
      "email": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "email",
        "reflect": false
      },
      "initials": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "initials",
        "reflect": false
      },
      "links": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "ModusNavbarProfileMenuLink[]",
          "resolved": "ModusNavbarProfileMenuLink[]",
          "references": {
            "ModusNavbarProfileMenuLink": {
              "location": "import",
              "path": "../modus-navbar.models"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        }
      },
      "reverse": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "reverse",
        "reflect": false
      },
      "username": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "username",
        "reflect": false
      },
      "variant": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'default' | 'blue'",
          "resolved": "\"blue\" | \"default\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "variant",
        "reflect": false,
        "defaultValue": "'default'"
      }
    };
  }
  static get events() {
    return [{
        "method": "linkClick",
        "name": "linkClick",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        }
      }, {
        "method": "signOutClick",
        "name": "signOutClick",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "MouseEvent",
          "resolved": "MouseEvent",
          "references": {
            "MouseEvent": {
              "location": "global"
            }
          }
        }
      }];
  }
}
