// @trimble-oss/modus-web-components/dist/collection/components/modus-navbar/notifications-menu/modus-navbar-notifications-menu.js
// eslint-disable-next-line
import { h } from '@stencil/core';
export class ModusNavbarNotificationsMenu {
  constructor() {
    this.reverse = undefined;
  }
  render() {
    const direction = this.reverse ? 'reverse' : '';
    return (h("div", { class: `notifications-menu ${direction}`, onClick: (event) => event.preventDefault() }, h("div", { class: "title" }, "Notifications"), h("slot", null)));
  }
  static get is() { return "modus-navbar-notifications-menu"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["modus-navbar-notifications-menu.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-navbar-notifications-menu.css"]
    };
  }
  static get properties() {
    return {
      "reverse": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "reverse",
        "reflect": false
      }
    };
  }
}
