// @trimble-oss/modus-web-components/dist/collection/components/modus-card/modus-card.js
// eslint-disable-next-line
import { h } from '@stencil/core';
export class ModusCard {
  constructor() {
    this.ariaLabel = undefined;
    this.height = '269px';
    this.width = '240px';
    this.borderRadius = undefined;
    this.showCardBorder = true;
    this.showShadowOnHover = true;
  }
  render() {
    return (h("article", { class: `${this.showShadowOnHover ? 'shadow' : ''} ${this.showCardBorder ? 'card-border' : ''}  `, "aria-label": this.ariaLabel, style: {
        height: this.height,
        width: this.width,
        'border-radius': this.borderRadius,
      } }, h("slot", null)));
  }
  static get is() { return "modus-card"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["modus-card.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-card.css"]
    };
  }
  static get properties() {
    return {
      "ariaLabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string | null",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The card's aria-label."
        },
        "attribute": "aria-label",
        "reflect": false
      },
      "height": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The height of the card."
        },
        "attribute": "height",
        "reflect": false,
        "defaultValue": "'269px'"
      },
      "width": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The width of the card."
        },
        "attribute": "width",
        "reflect": false,
        "defaultValue": "'240px'"
      },
      "borderRadius": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The border radius of the card."
        },
        "attribute": "border-radius",
        "reflect": false
      },
      "showCardBorder": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) A flag that controls the display of border."
        },
        "attribute": "show-card-border",
        "reflect": false,
        "defaultValue": "true"
      },
      "showShadowOnHover": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) A flag that controls the display of shadow box when the element is hovered."
        },
        "attribute": "show-shadow-on-hover",
        "reflect": false,
        "defaultValue": "true"
      }
    };
  }
}
