// @trimble-oss/modus-web-components/dist/collection/components/modus-table/parts/panel/modus-table-toolbar/modus-table-toolbar.js
import { Host, h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from '@stencil/core';
export class ModusTablePanel {
  constructor() {
    this.table = undefined;
    this.options = undefined;
  }
  render() {
    return (h(Host, null, h("div", { class: "table-toolbar" }, h("div", { class: "section" }, h("slot", { name: "group-left" })), h("div", { class: "section" }, h("slot", { name: "group-right" }), h("modus-table-dropdown-menu", { table: this.table, options: this.options })))));
  }
  static get is() { return "modus-table-toolbar"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["./modus-table-toolbar.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-table-toolbar.css"]
    };
  }
  static get properties() {
    return {
      "table": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "Table<unknown>",
          "resolved": "Table<unknown>",
          "references": {
            "Table": {
              "location": "import",
              "path": "@tanstack/table-core"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Table data."
        }
      },
      "options": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "ModusTableToolbarOptions",
          "resolved": "ModusTableToolbarOptions",
          "references": {
            "ModusTableToolbarOptions": {
              "location": "import",
              "path": "../../../models/modus-table.models"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(Optional) Table Panel options."
        }
      }
    };
  }
}
