// @trimble-oss/modus-web-components/dist/collection/components/modus-navbar/button-menu/modus-navbar-button-menu.js
import { h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from '@stencil/core';
export class ModusNavbarButtonMenu {
  constructor() {
    this.reverse = undefined;
  }
  render() {
    const menuClasses = {
      'button-menu': true,
      reverse: this.reverse,
    };
    return (h("div", { class: menuClasses }, h("slot", null)));
  }
  static get is() { return "modus-navbar-button-menu"; }
  static get originalStyleUrls() {
    return {
      "$": ["modus-navbar-button-menu.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-navbar-button-menu.css"]
    };
  }
  static get properties() {
    return {
      "reverse": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "reverse",
        "reflect": false
      }
    };
  }
}
