// @trimble-oss/modus-web-components/dist/collection/components/modus-accordion/modus-accordion.js
// eslint-disable-next-line
import { h } from '@stencil/core';
export class ModusAccordion {
  constructor() {
    this.ariaLabel = undefined;
  }
  render() {
    return (h("div", { "aria-label": this.ariaLabel, class: "accordion" }, h("slot", null)));
  }
  static get is() { return "modus-accordion"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["modus-accordion.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-accordion.css"]
    };
  }
  static get properties() {
    return {
      "ariaLabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string | null",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The accordion's aria-label."
        },
        "attribute": "aria-label",
        "reflect": false
      }
    };
  }
}
