// @trimble-oss/modus-web-components/dist/cjs/modus-table.constants-43b34557.js
'use strict';

const SORTED_ASCENDING = 'Sorted Ascending';
const SORTED_DESCENDING = 'Sorted Descending';
const SORT_ASCENDING = 'Sort Ascending';
const SORT_DESCENDING = 'Sort Descending';
const COLUMN_DEF_DATATYPE_KEY = 'dataType';
const COLUMN_DEF_SHOWTOTAL = 'showTotal';
const COLUMN_DEF_ROW_SELECTION_ID = 'row-checkbox';
const COLUMN_DEF_ROW_SELECTION_CSS = 'row-checkbox';
const COLUMN_DEF_DATATYPE_INTEGER = 'integer';
const COLUMN_DEF_DATATYPE_LINK = 'link';
const PAGINATION_DEFAULT_SIZES = [10, 20, 30];
const PAGINATION_PAGEVIEW_TEXT = 'Page View';
const PAGINATION_SUMMARY_TEXT = 'Showing result';
// Keyboard key constants
const KEYBOARD_ENTER = 'enter';
const KEYBOARD_TAB = 'tab';
const KEYBOARD_LEFT = 'arrowleft';
const KEYBOARD_RIGHT = 'arrowright';
const KEYBOARD_UP = 'arrowup';
const KEYBOARD_DOWN = 'arrowdown';
const KEYBOARD_SPACE = ' ';
const KEYBOARD_ESCAPE = 'escape';

exports.COLUMN_DEF_DATATYPE_INTEGER = COLUMN_DEF_DATATYPE_INTEGER;
exports.COLUMN_DEF_DATATYPE_KEY = COLUMN_DEF_DATATYPE_KEY;
exports.COLUMN_DEF_DATATYPE_LINK = COLUMN_DEF_DATATYPE_LINK;
exports.COLUMN_DEF_ROW_SELECTION_CSS = COLUMN_DEF_ROW_SELECTION_CSS;
exports.COLUMN_DEF_ROW_SELECTION_ID = COLUMN_DEF_ROW_SELECTION_ID;
exports.COLUMN_DEF_SHOWTOTAL = COLUMN_DEF_SHOWTOTAL;
exports.KEYBOARD_DOWN = KEYBOARD_DOWN;
exports.KEYBOARD_ENTER = KEYBOARD_ENTER;
exports.KEYBOARD_ESCAPE = KEYBOARD_ESCAPE;
exports.KEYBOARD_LEFT = KEYBOARD_LEFT;
exports.KEYBOARD_RIGHT = KEYBOARD_RIGHT;
exports.KEYBOARD_SPACE = KEYBOARD_SPACE;
exports.KEYBOARD_TAB = KEYBOARD_TAB;
exports.KEYBOARD_UP = KEYBOARD_UP;
exports.PAGINATION_DEFAULT_SIZES = PAGINATION_DEFAULT_SIZES;
exports.PAGINATION_PAGEVIEW_TEXT = PAGINATION_PAGEVIEW_TEXT;
exports.PAGINATION_SUMMARY_TEXT = PAGINATION_SUMMARY_TEXT;
exports.SORTED_ASCENDING = SORTED_ASCENDING;
exports.SORTED_DESCENDING = SORTED_DESCENDING;
exports.SORT_ASCENDING = SORT_ASCENDING;
exports.SORT_DESCENDING = SORT_DESCENDING;
