// @trimble-oss/modus-web-components/dist/cjs/modus-progress-bar.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-77fbe095.js');

const modusProgressBarCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.modus-progress-bar{display:flex;font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";height:1rem;outline:0.0625rem solid var(--modus-progress-border-color, #6a6e79);width:100%}.modus-progress-bar.default-background-color{background-color:var(--modus-progress-bg, #fff)}.modus-progress-bar .progress{display:flex;font-size:0.75rem;justify-content:center;overflow:hidden;transition:width 0.6s ease}.modus-progress-bar .progress.default-color{background-color:var(--modus-progress-bar-bg, #0063a3)}.modus-progress-bar .progress.default-text-color{color:#fff}.modus-progress-bar.small{height:0.5rem}.modus-progress-bar.compact{height:0.25rem}";

const ModusProgressBar = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.classBySize = new Map([
      ['default', 'default'],
      ['small', 'small'],
      ['compact', 'compact'],
    ]);
    this.ariaLabel = undefined;
    this.backgroundColor = undefined;
    this.color = undefined;
    this.maxValue = 100;
    this.minValue = 0;
    this.size = 'default';
    this.text = undefined;
    this.textColor = undefined;
    this.value = 0;
  }
  getProgressStyle(percentage) {
    const progressStyle = {
      backgroundColor: this.color,
      color: this.textColor,
      width: `${percentage}%`,
    };
    return progressStyle;
  }
  getProgressBarStyle() {
    const progressBarStyle = {
      backgroundColor: this.backgroundColor,
    };
    return progressBarStyle;
  }
  render() {
    const percentage = ((this.value - this.minValue) / (this.maxValue - this.minValue)) * 100;
    const progressBarBackgroundColorClass = this.backgroundColor ? '' : 'default-background-color';
    const progressColorClass = this.color ? '' : 'default-color';
    const progressTextColor = this.textColor ? '' : 'default-text-color';
    const progressBarClass = `
      modus-progress-bar
      ${progressBarBackgroundColorClass}
      ${this.classBySize.get(this.size)}
     `;
    const progressClass = `progress ${progressColorClass} ${progressTextColor}`;
    return (index.h(index.Host, { "aria-label": this.ariaLabel, "aria-valuemax": this.maxValue, "aria-valuemin": this.minValue, "aria-valuenow": this.value, role: "progressbar" }, index.h("div", { class: progressBarClass, style: this.getProgressBarStyle() }, index.h("div", { class: progressClass, style: this.getProgressStyle(percentage) }, this.size === 'default' && this.text))));
  }
};
ModusProgressBar.style = modusProgressBarCss;

exports.modus_progress_bar = ModusProgressBar;
