// @trimble-oss/modus-web-components/dist/esm/modus-table.constants-71fc7bdd.js
const SORTED_ASCENDING = 'Sorted Ascending';
const SORTED_DESCENDING = 'Sorted Descending';
const SORT_ASCENDING = 'Sort Ascending';
const SORT_DESCENDING = 'Sort Descending';
const COLUMN_DEF_DATATYPE_KEY = 'dataType';
const COLUMN_DEF_SHOWTOTAL = 'showTotal';
const COLUMN_DEF_ROW_SELECTION_ID = 'row-checkbox';
const COLUMN_DEF_ROW_SELECTION_CSS = 'row-checkbox';
const COLUMN_DEF_DATATYPE_INTEGER = 'integer';
const COLUMN_DEF_DATATYPE_LINK = 'link';
const PAGINATION_DEFAULT_SIZES = [10, 20, 30];
const PAGINATION_PAGEVIEW_TEXT = 'Page View';
const PAGINATION_SUMMARY_TEXT = 'Showing result';
// Keyboard key constants
const KEYBOARD_ENTER = 'enter';
const KEYBOARD_TAB = 'tab';
const KEYBOARD_LEFT = 'arrowleft';
const KEYBOARD_RIGHT = 'arrowright';
const KEYBOARD_UP = 'arrowup';
const KEYBOARD_DOWN = 'arrowdown';
const KEYBOARD_SPACE = ' ';
const KEYBOARD_ESCAPE = 'escape';

export { COLUMN_DEF_DATATYPE_LINK as C, KEYBOARD_ENTER as K, PAGINATION_SUMMARY_TEXT as P, SORTED_ASCENDING as S, COLUMN_DEF_DATATYPE_KEY as a, COLUMN_DEF_DATATYPE_INTEGER as b, KEYBOARD_SPACE as c, KEYBOARD_TAB as d, KEYBOARD_ESCAPE as e, KEYBOARD_LEFT as f, KEYBOARD_RIGHT as g, SORTED_DESCENDING as h, SORT_ASCENDING as i, SORT_DESCENDING as j, PAGINATION_PAGEVIEW_TEXT as k, COLUMN_DEF_ROW_SELECTION_ID as l, COLUMN_DEF_ROW_SELECTION_CSS as m, COLUMN_DEF_SHOWTOTAL as n, PAGINATION_DEFAULT_SIZES as o, KEYBOARD_DOWN as p, KEYBOARD_UP as q };
