// @trimble-oss/modus-web-components/dist/esm/modus-card.entry.js
import { r as registerInstance, h } from './index-3e825a40.js';

const modusCardCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");article{background-color:var(--modus-card-bg, #fff);border:0.0625rem solid;border-color:transparent;border-radius:0.125rem;color:var(--modus-card-color, #252a2e);font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";overflow:hidden;position:relative}article.shadow:hover{box-shadow:0 0 4px var(--modus-card-shadow-color, #3635454d)}article.card-border{border-color:var(--modus-card-border-color, #e0e1e9)}";

const ModusCard = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.ariaLabel = undefined;
    this.height = '269px';
    this.width = '240px';
    this.borderRadius = undefined;
    this.showCardBorder = true;
    this.showShadowOnHover = true;
  }
  render() {
    return (h("article", { class: `${this.showShadowOnHover ? 'shadow' : ''} ${this.showCardBorder ? 'card-border' : ''}  `, "aria-label": this.ariaLabel, style: {
        height: this.height,
        width: this.width,
        'border-radius': this.borderRadius,
      } }, h("slot", null)));
  }
};
ModusCard.style = modusCardCss;

export { ModusCard as modus_card };
