// @trimble-oss/modus-web-components/dist/collection/components/modus-tooltip/modus-tooltip.js
// eslint-disable-next-line
import { h } from '@stencil/core';
export class ModusTooltip {
  constructor() {
    this.ariaLabel = undefined;
    this.position = 'top';
    this.text = undefined;
    this.disabled = undefined;
  }
  render() {
    const className = `modus-tooltip ${this.position}`;
    const showTooltip = !this.disabled && this.text;
    return (h("div", { class: className }, h("slot", null), showTooltip && (h("div", { "aria-label": this.ariaLabel, class: 'text', role: "tooltip" }, this.text))));
  }
  static get is() { return "modus-tooltip"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["modus-tooltip.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["modus-tooltip.css"]
    };
  }
  static get properties() {
    return {
      "ariaLabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string | null",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The tooltip's aria-label."
        },
        "attribute": "aria-label",
        "reflect": false
      },
      "position": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'bottom' | 'left' | 'right' | 'top'",
          "resolved": "\"bottom\" | \"left\" | \"right\" | \"top\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "(optional) The tooltip's position relative to its content."
        },
        "attribute": "position",
        "reflect": false,
        "defaultValue": "'top'"
      },
      "text": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The tooltip's text."
        },
        "attribute": "text",
        "reflect": false
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Hide the tooltip"
        },
        "attribute": "disabled",
        "reflect": false
      }
    };
  }
}
